curvefs/src/client/rpcclient/metacache.cpp

```diff
@@ -183,6 +189,176 @@ bool MetaCache::IsLeaderMayChange(const CopysetGroupID &groupID) {
+
+    WriteLockGuard wl(rwlock4copysetInfoMap_);
+    copysetInfoMap_[key] = csinfo;
+}
+
+bool MetaCache::ListPartitions(uint32_t fsID) {
+    if (init_) {
+        return true;
+    }
+
+    WriteLockGuard wl4PartitionMap(rwlock4Partitions_);
+    WriteLockGuard wl4CopysetMap(rwlock4copysetInfoMap_);
+
+    if (init_) {
+        return true;
+    } else {
+        fsID_ = fsID;
+        LOG(INFO) << "init partition and copyset infos for {fsid:" << fsID
+                  << "}";
+    }
+
+    PatitionInfoList partitionInfos;
+    std::map<PoolIDCopysetID, CopysetInfo<MetaserverID>> copysetMap;
+    if (!DoListOrCreatePartitions(true, &partitionInfos, &copysetMap)) {
+        return false;
+    }
+
+    DoAddPartitionAndCopyset(partitionInfos, copysetMap);
+
+    init_ = true;
+
+    LOG(INFO) << "init partition and copyset infos for {fsid:" << fsID
+              << "} ok";
+    return true;
+}
+
+bool MetaCache::CreatePartitions(int currentNum,
+                                 PatitionInfoList *newPartitions) {
+    std::lock_guard<Mutex> lg(createMutex_);
+
+    // already create
+    {
+        ReadLockGuard rl(rwlock4Partitions_);
+        if (partitionInfos_.size() > currentNum) {
+            return true;
+        }
+    }
+
+    // create partition
+    PatitionInfoList partitionInfos;
+    std::map<PoolIDCopysetID, CopysetInfo<MetaserverID>> copysetMap;
+    if (!DoListOrCreatePartitions(false, &partitionInfos, &copysetMap)) {
+        return false;
+    }
+
+    // add partition and copyset info
+    WriteLockGuard wl4PartitionMap(rwlock4Partitions_);
+    WriteLockGuard wl4CopysetMap(rwlock4copysetInfoMap_);
+    DoAddPartitionAndCopyset(partitionInfos, copysetMap);
+}
+
+bool MetaCache::DoListOrCreatePartitions(
+    bool list, PatitionInfoList *partitionInfos,
+    std::map<PoolIDCopysetID, CopysetInfo<MetaserverID>> *copysetMap) {
+    if (list) {
+        if (!mdsClient_->ListPartition(fsID_, partitionInfos)) {
+            LOG(ERROR) << "list parition for {fsid:" << fsID_ << "} fail";
+            return false;
+        }
+    } else {
+        if (!mdsClient_->CreatePartition(
+                fsID_, metacacheopt_.createPartitionOnce, partitionInfos)) {
+            LOG(ERROR) << "create partition for fsid:" << fsID_ << " fail";
+            return false;
+        }
+    }
+
+    // gather partitionIDs
+    std::vector<PartitionID> partitionIDLsit;
+    for_each(partitionInfos->begin(), partitionInfos->end(),
+             [&](const PartitionInfo &info) {
+                 auto key = std::move(CalcLogicPoolCopysetID(
+                     CopysetGroupID(info.poolid(), info.copysetid())));
+                 bool exist = false;
+
+                 if (!list) {
+                     ReadLockGuard rl(rwlock4copysetInfoMap_);
+                     exist = (copysetInfoMap_.count(key) > 0);
+                 }
+
+                 if (!exist) {
+                     partitionIDLsit.emplace_back(info.partitionid());
```

Comment:
if exist copyset in copysetInfoMap_, then  mdsClient_->CreatePartition has no effect?